cpp/src/comms/mpi/comms_mpi.cpp

```diff
@@ -0,0 +1,291 @@
+      }
+    }
+  }
+  else if(std::is_same<value_t, float>::value) {
+    return ncclFloat32;
+  }
+  else if(std::is_same<value_t, double>::value) {
+    return ncclFloat64;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+constexpr MPI_Op get_mpi_reduce_op(ReduceOp reduce_op) {
+  if (reduce_op == ReduceOp::SUM) {
+    return MPI_SUM;
+  }
+  else if (reduce_op == ReduceOp::MAX) {
+    return MPI_MAX;
+  }
+  else if (reduce_op == ReduceOp::MIN) {
+    return MPI_MIN;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+constexpr ncclRedOp_t get_nccl_reduce_op(ReduceOp reduce_op) {
+  if (reduce_op == ReduceOp::SUM) {
+    return ncclSum;
+  }
+  else if (reduce_op == ReduceOp::MAX) {
+    return ncclMax;
+  }
+  else if (reduce_op == ReduceOp::MIN) {
+    return ncclMin;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+#endif
+
+Comm::Comm(int p) : _p{p} {
+#if ENABLE_OPG
+  // MPI
+  int flag{}, mpi_world_size;
+
+  MPI_TRY(MPI_Initialized(&flag));
+
+  if (flag == false) {
+    int provided{};
+    MPI_TRY(MPI_Init_thread(nullptr, nullptr, MPI_THREAD_MULTIPLE, &provided));
+    if (provided != MPI_THREAD_MULTIPLE) {
+      MPI_TRY(MPI_ERR_OTHER);
+    }
+    _finalize_mpi = true;
+  }
+
+  MPI_TRY(MPI_Comm_rank(MPI_COMM_WORLD, &_rank));
+  MPI_TRY(MPI_Comm_size(MPI_COMM_WORLD, &mpi_world_size));
+  CUGRAPH_EXPECTS( (_p == mpi_world_size), 
+                   "Invalid input arguments: p should match the number of MPI processes.");
+
+  _mpi_comm = MPI_COMM_WORLD;
+
+  // CUDA
+
+  CUDA_TRY(cudaGetDeviceCount(&_device_count));
+  _device_id = _rank % _device_count; // FixMe : assumes each node has the same number of GPUs
+  CUDA_TRY(cudaSetDevice(_device_id));
+
+  CUDA_TRY(
+    cudaDeviceGetAttribute(&_sm_count_per_device, cudaDevAttrMultiProcessorCount, _device_id));
+  CUDA_TRY(cudaDeviceGetAttribute(&_max_grid_dim_1D, cudaDevAttrMaxGridDimX, _device_id));
+  CUDA_TRY(cudaDeviceGetAttribute(&_max_block_dim_1D, cudaDevAttrMaxBlockDimX, _device_id));
+  CUDA_TRY(cudaDeviceGetAttribute(&_l2_cache_size, cudaDevAttrL2CacheSize, _device_id));
+  CUDA_TRY(
+    cudaDeviceGetAttribute(
+      &_shared_memory_size_per_sm, cudaDevAttrMaxSharedMemoryPerMultiprocessor, _device_id));
+
+  // NCCL
+
+  ncclUniqueId nccl_unique_id_p{};
+  if (get_rank() == 0) {
+    NCCL_TRY(ncclGetUniqueId(&nccl_unique_id_p));
+  }
+  MPI_TRY(MPI_Bcast(&nccl_unique_id_p, sizeof(ncclUniqueId), MPI_BYTE, 0, _mpi_comm));
+  NCCL_TRY(ncclCommInitRank(&_nccl_comm, get_p(), nccl_unique_id_p, get_rank()));
+  _finalize_nccl = true;
+#endif
+
+}
+
+#if ENABLE_OPG
+Comm::Comm(ncclComm_t comm, int size, int rank)
+  : _nccl_comm(comm), _p(size), _rank(rank) {
+
+  // CUDA
+  CUDA_TRY(cudaGetDeviceCount(&_device_count));
+  _device_id = _rank % _device_count; // FixMe : assumes each node has the same number of GPUs
+  CUDA_TRY(cudaSetDevice(_device_id)); // FixMe : check if this is needed or if python takes care of this
```

Comment:
Need to define what the FixMe tag looks like since you have various versions "Fix me" vs "FixMe"

Should be standard FIXME

Reply:
5b45a5c5